packages/react-async/src/hooks.ts

```diff
@@ -1,13 +1,121 @@
-import {useContext} from 'react';
+import {useState, useCallback, useContext} from 'react';
+import {Resolver} from '@shopify/async';
 import {useServerEffect} from '@shopify/react-effect';
+import {useMountedRef} from '@shopify/react-hooks';
+import {IfAllOptionalKeys} from '@shopify/useful-types';
+
 import {AsyncAssetContext} from './context/assets';
+import {AssetTiming, AsyncComponentType} from './types';
+
+export function usePreload<PreloadOptions extends object>(
+  ...args: IfAllOptionalKeys<
+    PreloadOptions,
+    [AsyncComponentType<any, any, PreloadOptions, any, any>, PreloadOptions?],
+    [AsyncComponentType<any, any, PreloadOptions, any, any>, PreloadOptions]
+  >
+): ReturnType<typeof args[0]['usePreload']> {
+  const [asyncComponent, options = {}] = args;
+  return (asyncComponent.usePreload as any)(options);
+}
+
+export function usePrefetch<PrefetchOptions extends object>(
+  ...args: IfAllOptionalKeys<
+    PrefetchOptions,
+    [AsyncComponentType<any, any, any, PrefetchOptions, any>, PrefetchOptions?],
+    [AsyncComponentType<any, any, any, PrefetchOptions, any>, PrefetchOptions]
+  >
+): ReturnType<typeof args[0]['usePrefetch']> {
+  const [asyncComponent, options = {}] = args;
+  return (asyncComponent.usePrefetch as any)(options);
+}
+
+export function useKeepFresh<KeepFreshOptions extends object>(
+  ...args: IfAllOptionalKeys<
+    KeepFreshOptions,
+    [
+      AsyncComponentType<any, any, any, any, KeepFreshOptions>,
+      KeepFreshOptions?
+    ],
+    [AsyncComponentType<any, any, any, any, KeepFreshOptions>, KeepFreshOptions]
+  >
+): ReturnType<typeof args[0]['useKeepFresh']> {
+  const [asyncComponent, options = {}] = args;
+  return (asyncComponent.useKeepFresh as any)(options);
+}
+
+interface Options {
+  immediate?: boolean;
+  assets?: AssetTiming;
+  styles?: AssetTiming;
+  scripts?: AssetTiming;
+}
+
+export function useAsync<T>(
+  resolver: Resolver<T>,
+  {assets, scripts = assets, styles = assets, immediate = true}: Options = {},
+) {
+  const [value, setValue] = useState<T | Error | null>(
+    () =>
+      immediate || typeof window !== 'undefined' ? resolver.resolved : null,
+  );
+
+  const mounted = useMountedRef();
+
+  const load = useCallback(
```

Comment:
I guess we have to `useCallback` for this stuff because we have `useEffect`s that depend on it?

Reply:
Indeed